<a href="https://colab.research.google.com/github/lorek/MethodsClassDimRed/blob/main/MoCaDR_List_nr_0_NumPy_intro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Methods of classification and dimensionality reduction**


Paweł Lorek  
University of Wrocław

<font face="Rage" size=2  > Updated: 18.02.2026 <font>

# LIST NR 0: Python and NumPy: introduction and exercises

# Introduction

In [2]:
import numpy as np
import time

# How to Use This List

This list is intended for independent work.  It is not meant to be followed step-by-step during classes, but rather to help you understand what you are expected to know and be comfortable with.

You should be able to solve most tasks on your own, using online resources when needed.



## Python: matrix, list

In [3]:
# "pure" Python way to represent matrix (here 2x3)
X = [[1, 2, 3], [4, 5, 6]]
print(X)
print(type(X))

[[1, 2, 3], [4, 5, 6]]
<class 'list'>


In [4]:
# ndarray
X = np.array([[1, 2, 3], [4, 5, 6]])
print(X)
print(type(X))

print(X)
X[0][0] = 50
print(X)
X[0,0] = 100 # == X[0][0]
print(X)

[[1 2 3]
 [4 5 6]]
<class 'numpy.ndarray'>
[[1 2 3]
 [4 5 6]]
[[50  2  3]
 [ 4  5  6]]
[[100   2   3]
 [  4   5   6]]


## Slicing

In [5]:
# For lists and strings, slicing creates a copy of the object
a = [1,2,3,4,5]
b = a[:] # a[start:end] <--- THIS IS SLICING

a[2] = -1
print(a)
print(b)

c = a   # <--- THIS IS NOT SLICING
a[2]=555
print(c)

[1, 2, -1, 4, 5]
[1, 2, 3, 4, 5]
[1, 2, 555, 4, 5]


In [6]:
X = np.array([[1, 2, 3, 4], [4, 5, 6, 7], [2, 2, 1, 1]])
print(X)



[[1 2 3 4]
 [4 5 6 7]
 [2 2 1 1]]


In [7]:
Y = X[:, :] # all rows and cols
print(Y)

[[1 2 3 4]
 [4 5 6 7]
 [2 2 1 1]]


In [8]:
Y = X[:2, :] # rows < 2, all cols
print(Y)

[[1 2 3 4]
 [4 5 6 7]]


In [9]:
Y = X[1:3, :] # rows 1,2
print(Y)

[[4 5 6 7]
 [2 2 1 1]]


In [10]:
Y = X[:2, 2:] # rows <2, cols>=2
print(Y)

[[3 4]
 [6 7]]


In [11]:
Y[0,0] = 1000
print(Y)
print(X)

[[1000    4]
 [   6    7]]
[[   1    2 1000    4]
 [   4    5    6    7]
 [   2    2    1    1]]


In [12]:
print(X.shape)
Y = X[1, :]
print(Y)
print(Y.shape)

Y = X[1:2, :]
print(Y)
print(Y.shape)

(3, 4)
[4 5 6 7]
(4,)
[[4 5 6 7]]
(1, 4)


## sum, mean, min, max etc

In [13]:
X = np.array([[1, 2, 3], [4, 5, 6]])
print(X)


[[1 2 3]
 [4 5 6]]


In [14]:
print(np.sum(X)) # sum all matrix
print(np.sum(X, axis=0)) # summing across columns
print(np.sum(X, axis=1)) # summing across rows

21
[5 7 9]
[ 6 15]


In [15]:
print(np.mean(X))
print(np.mean(X, axis=0))
print(np.mean(X, axis=1))

3.5
[2.5 3.5 4.5]
[2. 5.]


In [16]:
print(np.std(X))
print(np.std(X, axis=0))
print(np.std(X, axis=1))

1.707825127659933
[1.5 1.5 1.5]
[0.81649658 0.81649658]


In [17]:
print(np.min(X))
print(np.min(X, axis=0))
print(np.min(X, axis=1))

1
[1 2 3]
[1 4]


In [18]:
#element-wise minimum
x = np.array([1, 5, 3])
y = np.array([2, 2, 4])
print(np.minimum(x,y))

## ERROR: print(np.min(x,y))

[1 2 3]


## Creating arrays

In [19]:
X = np.zeros((3,2)) # matrix of size 3x2 with all entries 0
print(X)

[[0. 0.]
 [0. 0.]
 [0. 0.]]


In [20]:
X = np.ones((2,2)) # matrix of size 2x2 with all entries 1
print(X)

[[1. 1.]
 [1. 1.]]


In [21]:
X = np.full((3,2), 2026) # matrix of size 3x2 with all entries 2026
print(X)

[[2026 2026]
 [2026 2026]
 [2026 2026]]


In [22]:
x = np.arange(-np.pi, np.pi, 0.1) # points [-pi, pi+0.1, pi+0.2, ..., pi]
print(x)

[-3.14159265 -3.04159265 -2.94159265 -2.84159265 -2.74159265 -2.64159265
 -2.54159265 -2.44159265 -2.34159265 -2.24159265 -2.14159265 -2.04159265
 -1.94159265 -1.84159265 -1.74159265 -1.64159265 -1.54159265 -1.44159265
 -1.34159265 -1.24159265 -1.14159265 -1.04159265 -0.94159265 -0.84159265
 -0.74159265 -0.64159265 -0.54159265 -0.44159265 -0.34159265 -0.24159265
 -0.14159265 -0.04159265  0.05840735  0.15840735  0.25840735  0.35840735
  0.45840735  0.55840735  0.65840735  0.75840735  0.85840735  0.95840735
  1.05840735  1.15840735  1.25840735  1.35840735  1.45840735  1.55840735
  1.65840735  1.75840735  1.85840735  1.95840735  2.05840735  2.15840735
  2.25840735  2.35840735  2.45840735  2.55840735  2.65840735  2.75840735
  2.85840735  2.95840735  3.05840735]


In [23]:
x = np.linspace(0,1,9) # 9 points from [0,1] (including 0 and 1), equally distanced
print(x)


[0.    0.125 0.25  0.375 0.5   0.625 0.75  0.875 1.   ]


In [24]:
print(np.repeat(3, 4)) # repeat 4 times value 3
print(np.repeat(np.array([1,2,3]), 2))
print(np.tile(np.array([1,2,3]), 2)) # cf. with above

[3 3 3 3]
[1 1 2 2 3 3]
[1 2 3 1 2 3]


## Matrix operations

In [25]:
A = np.array([1, 2]) # vector 1D
B = np.array([[3], [4]]) # 2D matrix of size 2x1
print(A)
print(A.shape)
print(B)
print(B.shape)

[1 2]
(2,)
[[3]
 [4]]
(2, 1)


In [26]:
# Transpose:
print(A.T) # transpose of 1D vector = the same vector
print(A.T.shape)

[1 2]
(2,)


In [27]:
print(B.T)
print(B.T.shape)

[[3 4]]
(1, 2)


In [28]:
print(A[np.newaxis]) # 1D vector extended to 2D matrix of size 1x2
print(A[np.newaxis].shape)

[[1 2]]
(1, 2)


## Element-wise operatations
Most operataions are applied element-wise

In [29]:
X = np.array(range(0,10))
print(X)
print(X*2) ## *2
print(X**2) ## ^2

print(np.sin(X))
print(np.exp(x))

[0 1 2 3 4 5 6 7 8 9]
[ 0  2  4  6  8 10 12 14 16 18]
[ 0  1  4  9 16 25 36 49 64 81]
[ 0.          0.84147098  0.90929743  0.14112001 -0.7568025  -0.95892427
 -0.2794155   0.6569866   0.98935825  0.41211849]
[1.         1.13314845 1.28402542 1.45499141 1.64872127 1.86824596
 2.11700002 2.39887529 2.71828183]


## Indices

In [30]:
X = np.array(range(10,20))
print(X)

[10 11 12 13 14 15 16 17 18 19]


In [31]:
Y = X % 2
print(Y)

[0 1 0 1 0 1 0 1 0 1]


In [32]:
xs=np.sin(np.linspace(0,2*np.pi,20))
print("xs = ", xs)

xs_pos=xs[xs>0] # only those values which are >0

print("xs_pos = ",xs_pos)

print(np.where(xs>0))

xs =  [ 0.00000000e+00  3.24699469e-01  6.14212713e-01  8.37166478e-01
  9.69400266e-01  9.96584493e-01  9.15773327e-01  7.35723911e-01
  4.75947393e-01  1.64594590e-01 -1.64594590e-01 -4.75947393e-01
 -7.35723911e-01 -9.15773327e-01 -9.96584493e-01 -9.69400266e-01
 -8.37166478e-01 -6.14212713e-01 -3.24699469e-01 -2.44929360e-16]
xs_pos =  [0.32469947 0.61421271 0.83716648 0.96940027 0.99658449 0.91577333
 0.73572391 0.47594739 0.16459459]
(array([1, 2, 3, 4, 5, 6, 7, 8, 9]),)


In [33]:
xs[xs<0] = 0 # replace negative entries with zeros
print(xs)

[0.         0.32469947 0.61421271 0.83716648 0.96940027 0.99658449
 0.91577333 0.73572391 0.47594739 0.16459459 0.         0.
 0.         0.         0.         0.         0.         0.
 0.         0.        ]


In [34]:
# reshape:
X = np.arange(1,10)
print(X)

Y = X.reshape(3, 3)  # też widok!
print(Y)

print(X)

[1 2 3 4 5 6 7 8 9]
[[1 2 3]
 [4 5 6]
 [7 8 9]]
[1 2 3 4 5 6 7 8 9]


## Matrix Multiplication


In [35]:
# how it was for lists and strings:
print("123"*10)
a = [1,2,3]
print(a * 3)

print("*"*30)
print([x*3 for x in a])

123123123123123123123123123123
[1, 2, 3, 1, 2, 3, 1, 2, 3]
******************************
[3, 6, 9]


In [36]:
Y = np.arange(1,10).reshape((3,3))

print(Y)
print(Y * 3) # "ukryty" produkt Hadamarda
print(Y + 77)
print(Y * Y) # component-wise
print(Y @ Y) # matrix multiplication  (__matmul__)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[ 3  6  9]
 [12 15 18]
 [21 24 27]]
[[78 79 80]
 [81 82 83]
 [84 85 86]]
[[ 1  4  9]
 [16 25 36]
 [49 64 81]]
[[ 30  36  42]
 [ 66  81  96]
 [102 126 150]]


In [37]:
#compare:
X = np.array([[1,2,3],[4,5,6],[7,8,9]])
print(X,"\n")
Y = np.array([[-1,1,-1],[1,2,3],[3,2,1]])
print(Y)

print("\nX * Y = \n ", X*Y)
print("\nX @ Y = \n", X@Y)
print("\nEquivalent: np.dot(X,Y) = \n ", np.dot(X,Y))



[[1 2 3]
 [4 5 6]
 [7 8 9]] 

[[-1  1 -1]
 [ 1  2  3]
 [ 3  2  1]]

X * Y = 
  [[-1  2 -3]
 [ 4 10 18]
 [21 16  9]]

X @ Y = 
 [[10 11  8]
 [19 26 17]
 [28 41 26]]

Equivalent: np.dot(X,Y) = 
  [[10 11  8]
 [19 26 17]
 [28 41 26]]


## stacking

In [38]:
X = np.array([1, 2, 3])
Y = np.array([2, 3, 4])
print("X = ", X)
print("Y = ", Y)

print("np.vstack((X,Y)) = \n ", np.vstack((X,Y)))


print("np.hstack((X,Y)) = \n ", np.hstack((X,Y)))

X =  [1 2 3]
Y =  [2 3 4]
np.vstack((X,Y)) = 
  [[1 2 3]
 [2 3 4]]
np.hstack((X,Y)) = 
  [1 2 3 2 3 4]


## det and inverse

In [39]:
A = np.array([[0, 3, 2], [1, 0, 4], [0, 3, 6]])
print(A)
print(np.linalg.det(A))
B = np.linalg.inv(A)
print(B)

Y = np.array([3, -1, 2])
print(B @ Y)

[[0 3 2]
 [1 0 4]
 [0 3 6]]
-12.0
[[ 1.          1.         -1.        ]
 [ 0.5         0.         -0.16666667]
 [-0.25        0.          0.25      ]]
[ 0.          1.16666667 -0.25      ]


## Unique, set

In [40]:
data = [1, 1, 2, 2, 2, 3, 3, 67, -1]

# unique
unique = np.unique(data)
print(unique)

unique, counts = np.unique([1, 1, 2, 2, 2, 3, 3, 67, -1], return_counts=True)
print(unique, counts)


# set -- similar to unique, no guarantee on the order
unique_values = set(data)
print(unique_values)
print(type(unique_values))

[-1  1  2  3 67]
[-1  1  2  3 67] [1 2 3 2 1]
{1, 2, 67, 3, -1}
<class 'set'>


## Generating random numbers

In [41]:
# Generate 5 random U(0,1)
# OLD WAY:
x=np.random.uniform(0,1,5)
print(x)

[0.17129901 0.57776508 0.0541319  0.23997076 0.54230769]


⚠   `np.random.uniform` uses the legacy global random generator (MT19937).
This is considered outdated.

The recommended modern approach (since NumPy 1.17) is to use:



In [42]:
rng = np.random.default_rng()   # uses PCG64

x = rng.uniform(0,1,5)
print(x )

[0.63096483 0.47940254 0.28330788 0.05867565 0.12357293]


In [43]:
# setting seed for reproducibility
rng = np.random.default_rng(31415)   # uses PCG64 with seed 31415
x = rng.uniform(0,1,5)
print(x )
x = rng.uniform(0,1,5)
print(x )
rng = np.random.default_rng(31415)
x = rng.uniform(0,1,5)
print(x )

[0.49968933 0.11520254 0.86217723 0.83740069 0.20851741]
[0.98480099 0.70360289 0.24555866 0.8215201  0.21603391]
[0.49968933 0.11520254 0.86217723 0.83740069 0.20851741]


In [44]:
x = rng.choice([1,2,3,4,5,6], 1000, p=[0.1, 0.2, 0.3, 0.2, 0.1, 0.1]) # sample 1000 from [1,2,...,6] with distribution p
print(x)

[6 4 2 5 2 1 3 1 2 3 2 5 3 6 3 3 6 3 3 2 3 5 1 2 5 5 1 4 2 3 3 3 6 1 1 3 4
 3 4 2 2 6 5 1 1 6 3 4 5 2 2 1 3 4 5 2 3 4 3 6 2 2 3 3 6 5 1 2 2 2 4 3 3 2
 2 4 1 4 4 2 2 4 4 3 3 5 1 6 2 6 6 3 4 4 3 5 4 3 4 6 4 5 3 2 5 5 3 4 2 3 5
 3 6 6 3 5 3 3 1 6 3 2 2 3 3 3 4 4 2 2 6 4 3 3 3 2 6 6 2 3 5 3 4 4 5 3 2 1
 4 4 4 5 6 3 3 4 6 3 2 5 6 4 6 2 5 3 3 4 3 3 1 3 2 4 3 6 5 3 3 3 4 4 5 3 2
 2 4 3 4 4 6 6 5 2 3 4 1 3 3 5 2 1 3 2 3 1 2 5 4 2 3 3 3 3 5 3 3 1 3 3 3 6
 3 5 2 2 5 1 6 2 3 6 2 3 4 3 2 3 2 2 3 6 2 1 3 4 1 4 5 4 3 2 2 2 3 4 5 3 2
 6 5 6 4 1 3 3 1 1 5 3 4 1 5 3 1 3 5 4 3 6 1 3 1 3 6 3 2 2 4 3 3 3 3 5 2 1
 3 1 4 4 1 3 5 4 4 3 6 2 2 2 3 3 2 3 5 2 2 3 4 2 1 1 3 3 2 4 2 4 1 3 2 3 2
 6 1 5 6 4 6 4 6 3 3 6 4 4 3 1 6 3 4 4 2 3 4 3 3 2 6 1 2 2 2 3 5 5 3 2 1 3
 3 4 2 3 4 3 2 2 2 4 3 4 2 6 5 1 6 2 4 6 2 2 3 2 3 3 4 3 1 3 5 5 3 6 2 4 2
 4 4 3 1 3 2 5 3 6 6 6 4 1 3 3 2 5 4 3 3 1 5 5 1 3 3 3 3 4 2 5 4 2 6 3 2 3
 6 1 4 2 2 1 4 2 4 3 2 3 5 3 6 3 3 1 1 3 3 3 3 5 3 4 5 4 1 1 4 4 6 1 3 6 2
 4 3 4 4 1 1 3 1 4 2 3 6 

## Matplotlib

In [45]:
from matplotlib import pyplot as plt

In [46]:
%matplotlib inline ## <-- not needed in Colab, needed in older Jupiter Notebooks

UsageError: unrecognized arguments: ## <-- not needed in Colab, needed in older Jupiter Notebooks


In [ ]:
plt.hist(x, bins=6)
plt.show()

In [ ]:
X = rng.normal(1, 3, 10000) # N(1,3) distribution
plt.hist(X, bins=200)
plt.show()

In [ ]:
X = np.linspace(-2*np.pi, 2*np.pi, 1000)
plt.plot(X, np.sin(X))
plt.show()

In [ ]:
fig, axes = plt.subplots(1, 2, figsize=(15,3))
axes[0].plot(t, y1)
axes[1].plot(t, y2)

In [ ]:
t = np.arange(0, 2.5, 0.1)
y1 = np.sin(np.pi*t)
y2 = np.sin(np.pi*t + np.pi/2)
y3 = np.sin(np.pi*t - np.pi/2)
plt.plot(t, y1, 'b--', t, y2, 'g', t, y3, 'r-')
plt.show()

## Images

In [ ]:
import imageio.v3 as iio

url = "https://raw.githubusercontent.com/lorek/MethodsClassDimRed/refs/heads/main/pics/baboon.png"
image = iio.imread(url)

print(image.shape)

In [ ]:
plt.imshow(image)

In [ ]:
f, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(15,15))
ax1.imshow(image[:,:,0], cmap='Reds')
ax1.set_title('R')
ax2.imshow(image[:,:,1], cmap='Greens')
ax2.set_title('G')
ax3.imshow(image[:,:,2], cmap='Blues')
ax3.set_title('B')

In [ ]:
image2 = image.copy()

h,w,nr_colors= image2.shape

image2[0:(h//2),0:(w//2),1] = 0 # set color green (channel 1) to 0 in upper-left quarter of the image

plt.imshow(image2)

In [ ]:
image2.shape

In [ ]:
image_gray = image.mean(axis=2)
plt.imshow(image_gray, cmap='gray')

In [ ]:
random_image = np.random.rand(100, 100, 3)

## Estimating pi -- example

In [ ]:
R=10000

U=rng.uniform(0,1,R)
V=rng.uniform(0,1,R)
plt.scatter(U,V,s=2)

Compute $(X_1,\ldots,X_R)$, where $X_i=\mathbf{1}(U^2+V^2\leq1)$. Then we estimate $\pi$ by $\hat{Y}_R=4\cdot {1\over R}\sum_{i=1}^R {X_i}$

In [ ]:
X = (U**2 + V**2 <= 1).astype(int)


plt.figure(figsize=(6,6))

# quarter of circle
xx = np.arange(0, 1, 1/100)
plt.plot(xx, np.sqrt(1 - xx**2))

clrs = ['green', 'red', 'brown', 'orange']

for i in [0, 1]:
    plt.scatter(U[X==i], V[X==i], s=2,
                label=i, color=clrs[i])


Y_estim=np.mean(4*X)
print("Y_estim = ",Y_estim)


plt.legend()
plt.show()

# Some exercises

## E0.0 (warm-up)

Create:
* a=[1,2,...,100] (vector consiting of integer numbers from 1 to 100)
* b=[1,3,...,99]
* c=$[-\pi, -0.99\pi,...,0.99\pi, \pi]$
* A = matrix of size $10\times 10$:
$$\left[\begin{array}{llll}
1 & 2 & \ldots & 9 & 10\\
11 & 12 & \ldots & 19 & 20\\
\vdots \\
91 & 92 & \ldots & 99 & 100
\end{array}\right]$$
* B = upper-triangular matrix with ons on and above diagonal
* C = matrix of size $2\times 100$, where first row has elements $d_{1,i} = 1+2+\ldots + i$ and second $d_{2,i}=i!$
* D = matrix of size $100\times 100$ such that D[i,j]=1 if i divides j, 0 otherwise

## E0.1

1000 numbers for uniform distribution $U(-2,5)$

In [ ]:
rng = np.random.default_rng(0)
x = rng.uniform(-2,5, size=1000)

Standard devitaion of $n$ numbers $x_1,\ldots,x_n$
$$\sigma=\sqrt{{1\over n}\sum_{i=1}^n (x_i-\hat{x})^2}, \quad \textrm{where } \hat{x}={1\over n}\sum_{i=1}^n x_i$$
built-in function to compute it: `np.std(x)`

In [ ]:
print(np.std(x))

* a) compute `std` using only `np.mean`, `np.sqrt` and `**2`
* b) Make  a histogram of `x` with 50 bins. Compare it with density plot. Reapeat for 10k numbers
* c) Sample 1000 numbers from normal distribution with $\mu=3, \sigma^2=9$. Make their histogram with 50 bins. Compare with density plot. Repeat for 10k numbers
* d) Sample 1000 points $(x,y)$ where $x$ are sampled from $N(4,4)$ and $y$ are sampled from $N(5,1)$. Make a scatter plot of the points. Repeat for 10k points.
* e) Using data from d) estimate $P(X<Y)$ where X has distribution $N(4,4)$ and $Y$ has distribution $N(5,1)$. Can you provide exact answer? (computed based on probability/statistics)

## E0.2

In [ ]:
#Random vectors v,w,u of dimension d

d=100
v=rng.uniform(-10,10,size=d)
w=rng.uniform(-5,5,size=d)
u=rng.uniform(0,10,size=d)

n=200
#X = random (standard normal) matrix of size d x n
X=rng.normal(0,1,size=(d,n))

* a) compute lengths of vectors; weighted average of `v` weighted by weights `w`; dot product of `v` and `u`
* b) For $\mathbf{X}$ of size $d\times n$ (think: $n$ vectors from $\mathbb{R}^d$) compute:
* -- lengths of all $n$ vectors
* --  weighted average of all $n$ vectors weighted by weights `w` (result: vector of length $n$ with these averages)

## E0.3

For making plot use `matplotlib`

In [ ]:
import matplotlib.pyplot as plt

* a) make a 2d plot containing 10 points: $(3,5), (4,6), (7,7), (8,11), (10,12), (11,14), (12, 10), (16,15), (18, 19) , (20, 21)$
* b) add to the plot a line $y=0.6x+5$
* c) for each data $(x_i,y_i)$ point compute its distance to the line $y=0.6x+5$ defined simply as $dist((x_i,y_i), line)=(0.6x_i+5 - y_i)^2$
* d) try to  find $a,b$ such that sum of distances of all points from line $y=ax+b$ is smaller

## Important: avoid loops whenever possible!

### Example: compute $\sum_{i=0}^{10^6} \sin(i)$

In [ ]:
# Create a large array of numbers
numbers = np.arange(1_000_000)
numbers = np.arange(1000000)

# Time the Python loop approach
start_loop = time.time()
result_loop = 0;
for x in numbers:
  result_loop += np.sin(x)
end_loop = time.time()
time_loop = end_loop - start_loop

# Time the vectorized NumPy operation

start_vectorized = time.time()
result_vectorized = np.sum(np.sin(numbers))
end_vectorized = time.time()
time_vectorized = end_vectorized - start_vectorized

# Calculate speedup factor (avoid division by zero)
if time_vectorized > 0:
    speedup = time_loop / time_vectorized
else:
    speedup = float('inf')

print(f"result_loop = {result_loop:.6f} seconds, result_vectorized = {result_vectorized:.6f}")
print(f"Loop time: {time_loop:.6f} seconds")
print(f"Vectorized time: {time_vectorized:.6f} seconds")
print(f"Vectorized operation is {speedup:.2f} times faster than the loop.")


## E0.4 (bonus)

* Read about  [Monte Hall Problem](https://en.wikipedia.org/wiki/Monty_Hall_problem). Simulate the game show 10k times and estimate the probabilities of winning in using two strategies:
* do not change intially chosen door
* always change intially chosen door